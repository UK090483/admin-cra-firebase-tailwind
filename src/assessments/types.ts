import { Record, Identifier } from "../types";
export type AssessmentRating = 1 | 2 | 3 | 4;

export type AssessmentStatus = "created" | "hidden";

export interface IAssessmentPoints {
  isInnovative?: AssessmentRating;
  isClearUSP?: AssessmentRating;
  isTherePotential?: AssessmentRating;
  doTheyKnowCompetitors?: AssessmentRating;
  isBusinessModelSolid?: AssessmentRating;
  areNumbersPromising?: AssessmentRating;
  isFundingSolid?: AssessmentRating;
  isTeamExperienced?: AssessmentRating;
  willFutureCitiesBenefit?: AssessmentRating;
  isComplianceWithSDG?: AssessmentRating;
  isPresentationGood?: AssessmentRating;
}

export interface IAssessmentRecord extends Record, IAssessmentPoints {
  application_id: Identifier;
  application_name: string;
  judge_id: Identifier;
  judge_name: string;
  isInnovativeReason?: string;
  isClearUSPReason?: string;
  isTherePotentialReason?: string;
  doTheyKnowCompetitorsReason?: string;
  isBusinessModelSolidReason?: string;
  areNumbersPromisingReason?: string;
  isFundingSolidReason?: string;
  isTeamExperiencedReason?: string;
  willFutureCitiesBenefitReason?: string;
  isComplianceWithSDGReason?: string;
  isPresentationGoodReason?: string;
  status: AssessmentStatus;
  sum?: number;
}
