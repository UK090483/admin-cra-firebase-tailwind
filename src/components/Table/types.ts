import * as React from "react";

export interface IColumn {
  field: string;
  use: string;
  use_in_search?: boolean;
  use_in_display?: boolean;
  width?: string;
  render?: (row: IRow) => JSX.Element | string | number;
}
export interface IRowData {
  [name: string]: any;
}

export interface IBulkActionsProps {
  bulkItems: IBulkState;
  data: IRow[];
  setBulkItems: (bulkItems: IBulkState) => void;
}

export interface ITableOptions {
  bulkActions?: (props: IBulkActionsProps) => JSX.Element;
  showFilter?: boolean;
  fixedFilter?: (
    filter: ITableFilterState | null,
    setFilter: (filter: ITableFilterState | null) => void,
    columns: IColumn[],
    rows: IRow[]
  ) => JSX.Element;
}

export interface ITableProps {
  name?: string;
  columns: IColumn[];
  rows: IRow[];
  per_page?: number;
  searchFields?: string[];
  onRowClick?: IRowClickFunction;
  options?: ITableOptions;
  rowStyle?: (row: IRow) => string | void;
}
export interface ITableSortState {
  field: string;
  direction: "desc" | "asc";
}

export interface IBulkState {
  [id: string]: boolean;
}

export type ITableFilter = [
  "===" | "!==" | ">" | "<",
  string | boolean | undefined
];
export interface ITableFilterState {
  [id: string]: ITableFilter[];
}

export interface IColumn {
  field: string;
  use: string;
  use_in_search?: boolean;
  use_in_display?: boolean;
  render?: (row: IRow) => JSX.Element | string | number;
}

export interface ITableColumns extends Array<IColumn> {}

export interface IRow {
  [key: string]: any;
}

export type IRowClickFunction = (data: any) => void;
export declare type RowArray = IRow[];
export declare type renderFunction = (
  row: IRow,
  col: IColumn,
  display_value: string
) => JSX.Element;
export interface Iprop {
  rows: IRow[];
  columns: IColumn[];
  row_render?: renderFunction;
  per_page?: number;
  no_content_text?: string;
  debounce_search?: number;
  table_header?: string;
  show_search?: boolean;
  onRowClick?: IRowClickFunction;
}
/**
 * Interface describing data after the active page has been set
 * @props pagination: Has the full paginated data for the rows prop
 * @props back_button_clickable: indicates if back button is clickable
 * @props forward_button_clickable: Dictates if the forward button is clickable
 * @props current_rows: The rows to be shown in the table(The current page rows)
 * @props active_page_number: The currently selected page number
 * @param {string} search_string: The Text used in the search field
 *
 */
export interface ItableState {
  pagination: ItableLinks;
  back_button_clickable: boolean;
  forward_button_clickable: boolean;
  current_rows: IRow[];
  active_page_number: number;
  search_string: string;
}
export interface IpageNum {
  [key: number]: {
    page_row_array: IRow[];
    back_button_clickable: boolean;
    forward_button_clickable: boolean;
    is_active: boolean;
  };
}
export interface ItableLinks {
  page_number_store: number[];
  page_map: IpageNum;
  all_rows: IRow[];
}
export interface InavComponProp {
  back_button_clickable: boolean;
  forward_button_clickable: boolean;
}
export interface IpageActiveSet {
  table_links: ItableLinks;
  back_button_clickable: boolean;
  forward_button_clickable: boolean;
  current_rows: IRow[];
  active_page_number: number;
}
export interface IpaginateProps extends ItableState {
  backButtonOnclick: (event: React.MouseEvent) => void;
  forwardButtonOnclick: (event: React.MouseEvent) => void;
  pageNumberClick: (event: React.MouseEvent, page_number: number) => void;
}
export declare type stringObj = {
  [key: string]: any;
};
