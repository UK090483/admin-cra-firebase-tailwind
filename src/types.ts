import { Timestamp } from "@firebase/firestore-types";

export type Identifier = string;

export type FirebaseTimestamp = Timestamp;

export type RoutingParam = {
  id: Identifier;
};

export interface RecordBase {
  id: Identifier;
  created_at: Timestamp;
  created_by: Identifier;
  updated_at: Timestamp;
  updated_by: Identifier;
}

export type stage =
  | "first_filter"
  | "assign_prejudges"
  | "first_evaluation"
  | "final_evaluation";
