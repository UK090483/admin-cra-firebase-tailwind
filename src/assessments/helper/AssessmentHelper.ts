import { IAssessmentRecord } from "assessments/types";
import { IApplicationRecord } from "../../applications/ApplicationTypes";
import { IJudgeData } from "judges/state/judgesReducer";
import { Fields } from "./fields";
import { IJudgeAverages } from "applications/state/applicationsReducer";

interface AssessmentField {
  label: string;
  type: string;
  source: string;
  weight?: number;
  shortLabel: string;

  fields?: AssessmentFields;
}

interface AssessmentFields {
  main: AssessmentField;
  more: AssessmentField[];
}

export type AssessmentTypeKey = keyof AssessmentType;

export interface AssessmentType {
  isInnovative: AssessmentField;
  isClearUSP: AssessmentField;
  isTherePotential: AssessmentField;
  doTheyKnowCompetitors: AssessmentField;
  isBusinessModelSolid: AssessmentField;
  areNumbersPromising: AssessmentField;
  isFundingSolid: AssessmentField;
  isTeamExperienced: AssessmentField;
  willFutureCitiesBenefit: AssessmentField;
  isComplianceWithSDGs: AssessmentField;
  isPresentationGood: AssessmentField;
  // status: AssessmentField;
  // sum: AssessmentField;
}

// export type AssessmentType = {
//   [name in keyof IAssessmentRecord]: AssessmentField;
// };

type Entries<AssessmentType> = {
  [K in keyof AssessmentType]: [K, AssessmentType[K]];
}[keyof AssessmentType][];

function entries<AssessmentType>(obj: AssessmentType): Entries<AssessmentType> {
  return Object.entries(obj) as any;
}

export type AssessmentKey = keyof AssessmentType;

export class AssessmentC {
  fields: AssessmentType;
  constructor(fields: AssessmentType) {
    this.fields = fields;
  }

  // getMainField(name: AssessmentKey) {
  //   return this.fields[name].fields;
  // }

  // getLabel(name: AssessmentKey) {
  //   return this.fields[name].label;
  // }

  getQuestions() {
    const res = [];
    for (const [key, value] of entries(this.fields)) {
      // if (key !== "status" && key !== "sum") {
      res.push(this.fields[key]);
      // }
    }
    return res;
  }

  getQuestionsDetails() {
    const res: any = {};
    for (const [key, value] of entries(this.fields)) {
      // if (key !== "status" && key !== "sum") {
      res[key] = this.fields[key];
      // }
    }
    return res;
  }

  getTableFields() {
    const res = [];
    for (const [key, value] of entries(this.fields)) {
      res.push(this.fields[key]);
    }
    return res;
  }

  sumAssessmentPoints(assessment: IAssessmentRecord) {
    let res = 0;
    let isOk = true;
    for (let p of this.getQuestions()) {
      const value = assessment[p.source as keyof IAssessmentRecord];
      if (value && typeof value === "number") {
        res = res + value * (p.weight || 1);
      } else {
        isOk = false;
        break;
      }
    }
    return isOk ? res : undefined;
  }

  sumAssessmentPointsForJudgeApp(assessment: IAssessmentRecord) {
    let res = 0;
    for (let p of this.getQuestions()) {
      const value = assessment[p.source as keyof IAssessmentRecord];
      if (value && typeof value === "number") {
        res += value;
      }
    }
    return res;
  }

  getAvarage(assessment: IAssessmentRecord) {
    let res = 0;
    let count = 0;
    for (let p of this.getQuestions()) {
      const value = assessment[p.source as keyof IAssessmentRecord];
      if (value && typeof value === "number") {
        res = res + value;
        count++;
      }
    }

    return count === 11 ? res / count : undefined;
  }

  getAssessmentState(assessment: IAssessmentRecord) {
    let hasValues = false;
    let hasOpenValues = false;
    for (let p of this.getQuestions()) {
      const value = assessment[p.source as keyof IAssessmentRecord];
      if (value && typeof value === "number") {
        hasValues = true;
      } else {
        hasOpenValues = true;
      }
    }
    return hasValues ? (hasOpenValues ? "processed" : "completed") : "assigned";
  }

  evaluateAssessment(
    assessment: IAssessmentRecord,
    sumIn100?: boolean,
    integrateJudgeAverages?: boolean,
    judgeAvreges?: IJudgeAverages
  ) {
    const sum = this.sumAssessmentPoints(assessment);
    if (sum) {
      const withIntegration =
        integrateJudgeAverages &&
        judgeAvreges &&
        judgeAvreges[assessment.judge_id]
          ? sum * (1 - (judgeAvreges[assessment.judge_id] - 2) / 2)
          : sum;
      // console.log("----");
      // judgeAvreges &&
      //   judgeAvreges[assessment.judge_id] &&
      //   console.log(judgeAvreges[assessment.judge_id]);
      // judgeAvreges &&
      //   judgeAvreges[assessment.judge_id] &&
      //   console.log(1 - (judgeAvreges[assessment.judge_id] - 2) / 2);

      // console.log(withIntegration);
      // console.log(sum);

      const evaluatedSum = sumIn100
        ? (withIntegration / 64) * 100
        : withIntegration;
      return { ...assessment, sum: evaluatedSum } as IAssessmentRecord;
    }
    return assessment;
  }

  evaluateTableData(
    applications: IApplicationRecord[],
    judges: IJudgeData,
    sumIn100?: boolean,
    integrateJudgeAverages?: boolean,
    judgeAverages?: IJudgeAverages
  ) {
    return applications.map((application) => {
      const { sum, main, pre } = this.getSumApplication(
        application,
        judges,
        (sumIn100 = sumIn100 ? true : false),
        (integrateJudgeAverages = integrateJudgeAverages ? true : false),
        (judgeAverages = judgeAverages ? judgeAverages : {})
      );
      return {
        ...application,
        sum,
        mainSum: main,
        preSum: pre,
      } as IApplicationRecord;
    });
  }

  private getAverage = () => {
    let count = 0;
    let sum = 0;

    return {
      add: (number: number) => {
        sum += number;
        count++;
      },
      getResult: () => {
        if (sum && count) {
          return Math.round((sum / count) * 100) / 100;
        }
        return undefined;
      },
    };
  };

  getSumApplication = (
    application: IApplicationRecord,
    judges: IJudgeData,
    sumIn100: boolean,
    integrateJudgeAverages: boolean,
    judgeAverages: IJudgeAverages
  ) => {
    const main = this.getAverage();
    const pre = this.getAverage();
    const acc = this.getAverage();

    if (!judges) {
      return {
        sum: undefined,
        pre: undefined,
        main: undefined,
      };
    }

    if (application.assessments) {
      Object.values(application.assessments).forEach((_assessment) => {
        if (_assessment.status === "hidden") {
          return;
        }

        const assessment = this.evaluateAssessment(
          _assessment,
          sumIn100,
          integrateJudgeAverages,
          judgeAverages
        );
        assessment.sum && acc.add(assessment.sum);
        if (judges[assessment.judge_id].judgeType === "main") {
          assessment.sum && main.add(assessment.sum);
        } else {
          assessment.sum && pre.add(assessment.sum);
        }
      });
    }

    return {
      sum: acc.getResult(),
      pre: pre.getResult(),
      main: main.getResult(),
    };
  };
}

const AssessmentHelper = new AssessmentC(Fields);

export { AssessmentHelper };
