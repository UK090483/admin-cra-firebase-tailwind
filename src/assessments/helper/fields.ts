import { AssessmentType } from "./AssessmentHelper";

const Fields: AssessmentType = {
  isInnovative: {
    label: "Is the product/service innovative",
    shortLabel: "Inn",
    weight: 3,
    type: "number",
    source: "isInnovative",
    fields: {
      main: {
        label: "Is the product/service innovative",
        shortLabel: "Inn",
        type: "number",
        source: "isInnovative",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "I-R",
          type: "text",
          source: "isInnovativeReason",
        },
      ],
    },
  },

  isClearUSP: {
    label: "Is there a clear USP ?",
    shortLabel: "USP",
    type: "number",
    source: "isClearUSP",
    weight: 1,
    fields: {
      main: {
        label: "isClearUSP",
        shortLabel: "USP",
        type: "number",
        source: "isClearUSP",
      },
      more: [
        {
          label: "isClearUSPReason",
          shortLabel: "C-R",
          type: "text",
          source: "isClearUSPReason",
        },
      ],
    },
  },

  isTherePotential: {
    label: "Is there relevant market potential ?",
    shortLabel: "Pot",
    type: "number",
    source: "isTherePotential",
    weight: 1,
    fields: {
      main: {
        label: "isTherePotential",
        shortLabel: "P",
        type: "number",
        source: "isTherePotential",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "P-R",
          type: "text",
          source: "isTherePotentialReason",
        },
      ],
    },
  },

  doTheyKnowCompetitors: {
    label: "Do they know their competitors ?",
    shortLabel: "Cmp",
    type: "number",
    source: "doTheyKnowCompetitors",
    weight: 1,
    fields: {
      main: {
        label: "doTheyKnowCompetitors",
        shortLabel: "Cmp",
        type: "number",
        source: "doTheyKnowCompetitors",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "C-R",
          type: "text",
          source: "doTheyKnowCompetitorsReason",
        },
      ],
    },
  },

  isBusinessModelSolid: {
    label: "Is their business model solid and in balance with their resources?",
    shortLabel: "BMo",
    type: "number",
    source: "isBusinessModelSolid",
    weight: 1,
    fields: {
      main: {
        label: "isBusinessModelSolid",
        shortLabel: "BMo",
        type: "number",
        source: "isBusinessModelSolid",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "B-R",
          type: "text",
          source: "isBusinessModelSolidReason",
        },
      ],
    },
  },

  areNumbersPromising: {
    label: "Are the numbers promising?",
    shortLabel: "Num",
    type: "number",
    source: "areNumbersPromising",
    weight: 1,
    fields: {
      main: {
        label: "Are the numbers promising?",
        shortLabel: "Num",
        type: "number",
        source: "areNumbersPromising",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "P-R",
          type: "text",
          source: "areNumbersPromisingReason",
        },
      ],
    },
  },

  isFundingSolid: {
    label: "Is the funding solid?",
    shortLabel: "Fun",
    type: "number",
    source: "isFundingSolid",
    weight: 1,
    fields: {
      main: {
        label: "Is the funding solid?",
        shortLabel: "Fun",
        type: "number",
        source: "isFundingSolid",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "F-R",
          type: "text",
          source: "isFundingSolidReason",
        },
      ],
    },
  },

  isTeamExperienced: {
    label: "is Team Experienced",
    shortLabel: "Exp",
    type: "number",
    source: "isTeamExperienced",
    weight: 1,
    fields: {
      main: {
        label: "isTeamExperienced",
        shortLabel: "Exp",
        type: "number",
        source: "isTeamExperienced",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "E-R",
          type: "text",
          source: "isTeamExperiencedReason",
        },
      ],
    },
  },

  willFutureCitiesBenefit: {
    label: "Will our future cities benefit from their idea? ",
    shortLabel: "FCB",
    type: "number",
    source: "willFutureCitiesBenefit",
    weight: 3,
    fields: {
      main: {
        label: "willFutureCitiesBenefit",
        shortLabel: "FCB",
        type: "number",
        source: "willFutureCitiesBenefit",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "F-R",
          type: "text",
          source: "willFutureCitiesBenefitReason",
        },
      ],
    },
  },

  isComplianceWithSDG: {
    label: "Is the business in compliance with the SDGs?",
    shortLabel: "SDG",
    type: "number",
    source: "isComplianceWithSDG",
    fields: {
      main: {
        label: "isComplianceWithSDGs",
        shortLabel: "SDG",
        type: "number",
        source: "isComplianceWithSDGs",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "F-R",
          type: "text",
          source: "isComplianceWithSDGsReason",
        },
      ],
    },
  },
  isPresentationGood: {
    label:
      "Is the startup presentation based on elevator pitch, pitch deck and website good?",
    shortLabel: "PDF",
    type: "number",
    source: "isPresentationGood",
    weight: 2,
    fields: {
      main: {
        label: "isPresentationGood",
        shortLabel: "PDF",
        type: "number",
        source: "isPresentationGood",
      },
      more: [
        {
          label: "is innovative",
          shortLabel: "F-R",
          type: "text",
          source: "isInnovativeReason",
        },
      ],
    },
  },
};

export { Fields };
