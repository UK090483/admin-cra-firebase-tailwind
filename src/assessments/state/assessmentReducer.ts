import { AnyAction, createSlice, current } from "@reduxjs/toolkit";
import {} from "redux-firestore";
import { difference, isEqual } from "underscore";
import { IJudgeRecord, judgeType } from "../../judges/JudgeTypes";
import { AssessmentHelper } from "../helper/AssessmentHelper";
import { IAssessmentRecord } from "../types";

export interface IAssessmentState {
  data: { [k: string]: { [A: string]: IAssessmentRecord } } | null;
  ordered: any[];
  assessmentCount: number;
  judgeAverages: { [A: string]: number | undefined };
  sumByApplicationId: {
    [A: string]: { main: number; pre: number; all: number };
  };
  judgeTypes: { [k: string]: judgeType };
}

function isJudgeResponse(action: AnyAction): action is AnyAction {
  return (
    action.type === "@@reduxFirestore/LISTENER_RESPONSE" &&
    action.meta.collection === "judges"
  );
}

function isModifiedJudge(action: AnyAction): action is AnyAction {
  return (
    action.type === "@@reduxFirestore/DOCUMENT_MODIFIED" &&
    action.meta.collection === "judges"
  );
}

const initialState: IAssessmentState = {
  data: {},
  ordered: [],
  assessmentCount: 0,
  judgeAverages: {},
  sumByApplicationId: {},
  judgeTypes: {},
};

const AssessmentSlice = createSlice({
  name: "judgesApp",
  initialState: initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addMatcher(isJudgeResponse, (state, action) => {
        const collectSum = AssessmentHelper.sumMultiple();

        if (Array.isArray(action.payload.ordered)) {
          action.payload.ordered.forEach((judge: IJudgeRecord) => {
            if (!judge.assessments) return;
            const judgeAvr = AssessmentHelper.getAverage();

            Object.values(judge.assessments).forEach(
              (assessment: IAssessmentRecord) => {
                const sum = AssessmentHelper.sumAssessmentPoints(
                  assessment as IAssessmentRecord
                );

                if (sum) judgeAvr.add(sum);

                // state.ordered.push({ ...assessment, sum: sum });

                state.assessmentCount++;

                AddingAssessmentsByApplicationId(state, assessment, sum);

                collectSum.add(assessment.application_id, sum, judge.judgeType);
              }
            );

            state.judgeAverages[judge.id] = judgeAvr.getResult();
            state.sumByApplicationId = collectSum.getResult();
          });
        }
      })
      .addMatcher(isModifiedJudge, (state, action) => {
        const currentState = current(state).data;
        const _judge = action.payload.data as IJudgeRecord;
        if (_judge.assessments) {
          const judgeAvr = AssessmentHelper.getAverage();
          Object.values(_judge.assessments).forEach((assessment) => {
            const sum = AssessmentHelper.sumAssessmentPoints(assessment);

            if (sum) judgeAvr.add(sum);

            AddingAssessmentsByApplicationId(state, assessment, sum);

            // const newSum = updateSumNeeded(currentState, assessment);

            // if (newSum) {
            //   state.sumByApplicationId[assessment.application_id][
            //     _judge.judgeType
            //   ] = newSum;
            // }
          });
          state.judgeAverages[action.payload.id] = judgeAvr.getResult();
        }
      });
  },
});

export default AssessmentSlice.reducer;

const AddingAssessmentsByApplicationId = (
  state: any,
  assessment: IAssessmentRecord,
  sum: number | undefined
) => {
  if (!state.data[assessment.application_id])
    state.data[assessment.application_id] = {};

  state.data[assessment.application_id][assessment.judge_id] = {
    ...assessment,
    sum: sum,
  };
};

// const updateSumNeeded = (
//   current: any,
//   next: IAssessmentRecord
// ): number | null => {
//   const _currentAssessment =
//     current[next.application_id] && current[next.application_id][next.judge_id];
//   if (!_currentAssessment) return null;

//   if (AssessmentHelper.isEqual(_currentAssessment, next)) return null;

//   const avr = AssessmentHelper.getAverage();

//   Object.values(current[next.application_id]).forEach((assessment) => {
//     if (assessment && typeof assessment === "object") {
//       const _a = assessment as IAssessmentRecord;
//       _a.sum && avr.add(_a.sum);
//     }
//   });
//   return avr.getResult() || null;
//   console.log(next.application_id + " needs update");
// };
