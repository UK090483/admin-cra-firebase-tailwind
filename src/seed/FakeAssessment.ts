import * as faker from "faker";
import { IAssessmentRecord } from "assessments/types";
import { base } from "./helper";

const pointGenerator = () => {
  let sum = 0;
  const ran: (1 | 2 | 3 | 4)[] = [1, 2, 3, 4];
  return {
    point: (): 1 | 2 | 3 | 4 => {
      const ranPoint = ran[Math.floor(Math.random() * ran.length)];
      sum += ranPoint;
      return ranPoint;
    },
    sum: () => sum,
  };
};

const FakeAssessment: (
  application_id: string,
  judge_id: string,
  mock_in_process?: boolean
) => IAssessmentRecord = (
  application_id,
  judge_id,
  mock_in_process = false
) => {
  const { point } = pointGenerator();

  return {
    ...base,
    application_id: application_id,
    application_name: "",
    judge_id: judge_id,
    judge_name: "",
    ...((faker.random.boolean() || !mock_in_process) && {
      isInnovative: point(),
    }),
    isInnovativeReason: faker.lorem.paragraph(3),

    ...((faker.random.boolean() || !mock_in_process) && {
      isClearUSP: point(),
    }),
    isClearUSPReason: faker.lorem.paragraph(3),
    ...((faker.random.boolean() || !mock_in_process) && {
      isTherePotential: point(),
    }),
    isTherePotentialReason: faker.lorem.paragraph(3),
    doTheyKnowCompetitors: point(),
    doTheyKnowCompetitorsReason: faker.lorem.paragraph(3),
    isBusinessModelSolid: point(),
    isBusinessModelSolidReason: faker.lorem.paragraph(3),
    areNumbersPromising: point(),
    areNumbersPromisingReason: faker.lorem.paragraph(3),
    isFundingSolid: point(),
    isFundingSolidReason: faker.lorem.paragraph(3),
    isTeamExperienced: point(),
    isTeamExperiencedReason: faker.lorem.paragraph(3),
    willFutureCitiesBenefit: point(),
    willFutureCitiesBenefitReason: faker.lorem.paragraph(3),
    isComplianceWithSDG: point(),
    isComplianceWithSDGReason: faker.lorem.paragraph(3),
    isPresentationGood: point(),
    isPresentationGoodReason: faker.lorem.paragraph(3),
    status: "created",
  };
};

export default FakeAssessment;
