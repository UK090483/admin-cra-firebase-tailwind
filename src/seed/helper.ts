import { IApplicationAssessment } from "applications/ApplicationTypes";
import FakeAssessment from "./FakeAssessment";
import { IAssessmentRecord } from "assessments/types";
import { IJudgeRecord } from "judges/JudgeTypes";
import firebase from "firebase";

const getArray: (amount: number) => number[] = (amount) => {
  const all = new Array(amount);
  return all.fill(1).map((item, index) => index);
};

function getRandom(arr: number[], n: number): number[] {
  var result = new Array(n),
    len = arr.length,
    taken = new Array(len);
  if (n > len)
    throw new RangeError("getRandom: more elements taken than available");
  while (n--) {
    var x = Math.floor(Math.random() * len);
    result[n] = arr[x in taken ? taken[x] : x];
    taken[x] = --len in taken ? taken[len] : len;
  }
  return result;
}

export const base = {
  id: "",
  created_at: firebase.firestore.Timestamp.now(),
  updated_at: firebase.firestore.Timestamp.now(),
  updated_by: "",
  created_by: "",
};

const AssessmentGenerator = (judges: IJudgeRecord[], amount: number) => {
  let all: IAssessmentRecord[] = [];
  return {
    getApplicationAssessments: (id: string) => {
      const ranJudgesIndexes = judges.map((item, index) => index);
      const ranJudges = getRandom(ranJudgesIndexes, amount);
      const ranAssessments = ranJudges.map((item) =>
        FakeAssessment(id, "Fake" + "Judge" + item)
      );

      const lastIndex: number = all.length;
      all = [...all, ...ranAssessments];

      const assessResult: IApplicationAssessment = {};

      ranAssessments.forEach((item, index) => {
        const makeID = "Fake" + "Assessment" + (index + lastIndex);
        assessResult[makeID] = item;
      });

      return assessResult;
    },
    getAllAssessments: () => all,
  };
};
function chunk(array: any[], size: number) {
  const chunked_arr = [];
  for (let i = 0; i < array.length; i++) {
    const last = chunked_arr[chunked_arr.length - 1];
    if (!last || last.length === size) {
      chunked_arr.push([array[i]]);
    } else {
      last.push(array[i]);
    }
  }
  return chunked_arr;
}

export { AssessmentGenerator, getArray, chunk };
