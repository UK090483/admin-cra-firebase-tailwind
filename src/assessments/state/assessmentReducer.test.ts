/**
 * @jest-environment jsdom
 */

import FakeAssessment from "seed/FakeAssessment";
import assessmentReducer from "./assessmentReducer";
import { AssessmentHelper } from "../helper/AssessmentHelper";
import { IAssessmentRecord } from "../types";

const initialState = {
  data: {},
  ordered: [],
  assessmentCount: 0,
  judgeAverages: {},
  sumByApplicationId: {},
  judgeTypes: {},
};

const testAssessment = () => {
  return FakeAssessment("testApplication1", "testJudge1");
};

const ordered = [{}, {}, {}];

describe("AssessmentReducer", () => {
  it("should return Initial State", () => {
    expect(assessmentReducer(undefined, { type: "noType" })).toMatchSnapshot();
  });

  it("should return Initial State 2", () => {
    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: { ordered: {} },
        meta: { collection: "bla" },
      })
    ).toMatchSnapshot();
  });

  it("should return correct  assessmentCount", () => {
    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: { ordered: {} },
        meta: { collection: "judges" },
      }).assessmentCount
    ).toBe(0);

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: { ordered: [{}, {}, {}] },
        meta: { collection: "judges" },
      }).assessmentCount
    ).toBe(0);

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: { ordered: [{ assessments: {} }, {}, {}] },
        meta: { collection: "judges" },
      }).assessmentCount
    ).toBe(0);

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: {
          ordered: [
            { judgeType: "pre", assessments: { testAssessment1: {} } },
            {},
            {},
          ],
        },
        meta: { collection: "judges" },
      }).assessmentCount
    ).toBe(1);

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: {
          ordered: [
            { judgeType: "pre", assessments: { testAssessment1: {} } },
            { judgeType: "pre", assessments: { testAssessment2: {} } },
            {},
          ],
        },
        meta: { collection: "judges" },
      }).assessmentCount
    ).toBe(2);
  });

  it("should return judge Averages", () => {
    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: { ordered: [] },
        meta: { collection: "judges" },
      }).judgeAverages
    ).toStrictEqual({});

    const testAssessment1 = { ...testAssessment() };
    const testAssessment2 = { ...testAssessment() };
    const testAssessment3 = { ...testAssessment() };
    delete testAssessment3.isBusinessModelSolid;

    const sumTestAssessment1 = AssessmentHelper.sumAssessmentPoints(
      testAssessment1
    );
    const sumTestAssessment2 = AssessmentHelper.sumAssessmentPoints(
      testAssessment2
    );

    const expectedAvr =
      sumTestAssessment1 &&
      sumTestAssessment2 &&
      (sumTestAssessment1 + sumTestAssessment2) / 2;

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: {
          ordered: [
            {
              id: "testJudge1",
              judgeType: "pre",
              assessments: {
                testAssessment2: { ...testAssessment1 },
              },
            },
          ],
        },
        meta: { collection: "judges" },
      }).judgeAverages
    ).toStrictEqual({
      testJudge1: sumTestAssessment1,
    });

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: {
          ordered: [
            {
              id: "testJudge1",
              judgeType: "pre",
              assessments: {
                testAssessment1: { ...testAssessment1 },
                testAssessment2: { ...testAssessment2 },
              },
            },
          ],
        },
        meta: { collection: "judges" },
      }).judgeAverages
    ).toStrictEqual({
      testJudge1: expectedAvr,
    });

    expect(
      assessmentReducer(initialState, {
        type: "@@reduxFirestore/LISTENER_RESPONSE",
        payload: {
          ordered: [
            {
              id: "testJudge1",
              judgeType: "pre",
              assessments: {
                testAssessment1: { ...testAssessment1 },
                testAssessment2: { ...testAssessment2 },
                testAssessment3: { ...testAssessment3 },
              },
            },
          ],
        },
        meta: { collection: "judges" },
      }).judgeAverages
    ).toStrictEqual({
      testJudge1: expectedAvr,
    });
  });

  it("should return assessment Data", () => {});
});
