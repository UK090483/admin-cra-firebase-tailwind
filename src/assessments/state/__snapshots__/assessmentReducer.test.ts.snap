// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssessmentReducer should return Initial State 1`] = `
Object {
  "assessmentCount": 0,
  "data": Object {},
  "judgeAverages": Object {},
  "judgeTypes": Object {},
  "ordered": Array [],
  "sumByApplicationId": Object {},
}
`;

exports[`AssessmentReducer should return Initial State 2 1`] = `
Object {
  "assessmentCount": 0,
  "data": Object {},
  "judgeAverages": Object {},
  "judgeTypes": Object {},
  "ordered": Array [],
  "sumByApplicationId": Object {},
}
`;
